### ISSUE TITLE
Unexpected Koalas groupby.apply groupby.agg behavior

### ISSUE THREAD

**Issue by matthewgson** [databricks/koalas#1300]
I found some unexpected groupby.agg or groupby.apply behavior.
Here's a reproducible example.

```
test = pd.DataFrame({'secid': {0: 125047, 1: 125047, 2: 125047, 3: 125047, 4: 125047, 5: 125047},
 'date': {0: pd.Timestamp('2018-03-05 00:00:00'),
  1: pd.Timestamp('2018-03-05 00:00:00'),
  2: pd.Timestamp('2018-03-05 00:00:00'),
  3: pd.Timestamp('2018-03-05 00:00:00'),
  4: pd.Timestamp('2018-03-05 00:00:00'),
  5: pd.Timestamp('2018-03-05 00:00:00')},
 'days': {0: 182, 1: 182, 2: 182, 3: 182, 4: 182, 5: 182},
 'delta': {0: -20, 1: -15, 2: -10, 3: 10, 4: 15, 5: 20},
 'impl_volatility': {0: 0.251335,
  1: 0.321205,
  2: 0.433905,
  3: 0.327283,
  4: 0.307678,
  5: 0.28428},
'abs_delta': {0: 20, 1: 15, 2: 10, 3: 10, 4: 15, 5: 20}})


def my_divider(x) -> float:
    try:
        return x.iloc[0]/x.iloc[1]
    except IndexError: 
        pass


test
secid | date | days | delta | impl_volatility | abs_delta
-- | -- | -- | -- | -- | --
125047 | 2018-03-05 | 182 | -20 | 0.251335 | 20
125047 | 2018-03-05 | 182 | -15 | 0.321205 | 15
125047 | 2018-03-05 | 182 | -10 | 0.433905 | 10
125047 | 2018-03-05 | 182 | 10 | 0.327283 | 10
125047 | 2018-03-05 | 182 | 15 | 0.307678 | 15
125047 | 2018-03-05 | 182 | 20 | 0.284280 | 20


```
I was trying to group by multiple columns and calculate the ratio using ```my_divider``` function. 

When I run below codes in pandas, it works as expected.
```
test.groupby(['secid','date','days','abs_delta'])['impl_volatility'].apply(sum)
>>>
secid   date        days  abs_delta
125047  2018-03-05  182   10           0.761188
                          15           0.628883
                          20           0.535615
Name: impl_volatility, dtype: float64

test.groupby(['secid','date','days','abs_delta'])['impl_volatility'].apply(my_divider)
>>>secid   date        days  abs_delta
125047  2018-03-05  182   10           1.325779
                          15           1.043965
                          20           0.884111
Name: impl_volatility, dtype: float64

test.groupby(['secid','date','days','abs_delta']).agg(vol_ratio = ('impl_volatility',my_divider))
>>>
secid	date	days	abs_delta	vol_ratio
125047	2018-03-05	182	10	1.325779
                                                15	1.043965
                                                20	0.884111
```

However, in Koalas dataframe

```
test2 = ks.from_pandas(test)
test2.groupby(['secid','date','days','abs_delta'])['impl_volatility'].apply(sum)
>>>
0    0.535615
1    0.628883
2    0.761188
3    0.761188
4    0.628883
5    0.535615
# six outputs (three * 2 duplicates)


test2.groupby(['secid','date','days','abs_delta']).agg(mycolumn = ('impl_volatility','sum'))
>>>
secid	date	days	abs_delta	mycolumn
125047	2018-03-05	182	15	0.628883
                                                20	0.535615
                                                10	0.761188

```
.agg works with 'sum' function as expected but .agg does not understand function defined above.
```
test2.groupby(['secid','date','days','abs_delta']).agg(mycolumn = ('impl_volatility','my_divider'))
>>>
Py4JJavaError: An error occurred while calling o3520.agg.
: org.apache.spark.sql.AnalysisException: Undefined function: 'my_divider'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'
AnalysisException: "Undefined function: 'my_divider'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'.; line 1 pos 0"
```
So I tried registering function, but
```
from pyspark.sql.functions import pandas_udf, PandasUDFType
@pandas_udf('double') 
def my_divider_udf(x):
    try:
        return x.iloc[0]/x.iloc[1]
    except:
        pass

spark.udf.register('my_udf',my_divider_udf)

```
I get 
```

test2.groupby(['secid','date','days','abs_delta']).agg(pc_ratio=('impl_volatility','my_udf'))

Py4JJavaError: An error occurred while calling o2588.agg.
: org.apache.spark.sql.AnalysisException: expression '`impl_volatility`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;;
```

Where .apply throws different type of error.
```
test2.groupby(['secid','date','days','abs_delta'])['impl_volatility'].apply(my_divider)
>>>
Py4JJavaError: An error occurred while calling o1084.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 94 in stage 22.0 failed 1 times, most recent failure: Lost task 94.0 in stage 22.0 (TID 167, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 377, in main
    process()
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 372, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/serializers.py", line 290, in dump_stream
    for series in iterator:
  File "<string>", line 1, in <lambda>
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 113, in wrapped
    result = f(pd.concat(value_series, axis=1))
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/util.py", line 99, in wrapper
    return f(*args, **kwargs)
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/databricks/koalas/groupby.py", line 978, in rename_output
    pdf = func(pdf)
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/databricks/koalas/groupby.py", line 1628, in pandas_transform
    return pdf.transform(func)
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py", line 6735, in transform
    return super().transform(func, *args, **kwargs)
  File "/Users/matthewson/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py", line 10814, in transform
    raise ValueError("transforms cannot produce " "aggregated results")
ValueError: transforms cannot produce aggregated results

```

I'm not sure why I'm seeing last two errors in Koalas with my limited experience. I feel I might be doing something wrong here.

**Comment by HyukjinKwon** [comment]
@matthewgson, thanks for reporting an issue. It was a horrible mistake .. The fix is on the way

**Comment by HyukjinKwon** [comment]
Made a fix at #1339

**Comment by matthewgson** [comment]
@HyukjinKwon Thank you for updates! but one thing I'm curious about is `AnalysisException: "Undefined function: 'my_divider'  This function is neither a registered temporary function nor a permanent function registered in the database 'default'` part above. 
Will it be necessary to define & register a function using `@pandas_udf()` and `spark.udf.register('my_udf',my_divider_udf)` before applying it in Koalas dataframe in the future?
